### PR TITLE
[Merged by Bors] - checkpoint: preserve own atx chain

### LINKED ISSUE #4624: checkpoint: correctly preserve miners inflight atxs and its dependencies
## Description
checkpoint recovery on testnet-05 exposed the following problem:
- if a miner is still initializing the storage before the checkpoint, their atx will be rejected later if the commitment atx it used was not part of the snapshot.
- when a miner's new atx is not part of the snapshot (either its waiting for poet proof or just published), the preserved atx will not be accepted by peers if
  - the number of atxs per miners were not enough to capture the dependency chain of a miner's own atx
  - if the miner's atx depends on a golden atx of a previous checkpoint snapshot

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation
Closes #4624 

## Changes
code
- if miner doesn't have any atx, check post data dir for metadata and preserve the commitment atx (and its deps) used for initialization
- if miner has an atx that's not part of the snapshot, preserve its chain of deps as far as it can
- if miner has an pending atx in nipost metadata, also preserve its positioning atx
- abort preserving atx if any of its dependency is part of the snapshot of previous checkpoint
- atx dependencies are recursively discovered. the order is important for the validation path
- once node start services after recovering from checkpoint, validate and save the preserved atxs and poet proofs.

systest TestCheckpoint
- publish snapshot that only contains half of the miners atxs. the other miners need to preserve the chain of its own atxs
- using checkpoint url for testing (instead of local snapshot files)

cleanup
- remove recovery from local file
- move checkpoint related data to common/types, as post init dependency is introduced to checkpoint package.

note:
i tried shutting down nodes gracefully in admin::Recover() RPC instead of panicking during TestCheckpoint. but kubernetes doesn't restart the nodes automatically, which means the nodes will not restart to do the recovery.

### CODE REVIEW

**Comment by dshulyak** [comment]
what is the expected behavior when format has changed and user preserved own atx? will it crush its own node when it tries to encode atx? will it be disconnected from whole network because it will serve malformed data repeatedly?

File: checkpoint/recovery.go [hunk 1]
```diff
@@ -216,65 +219,149 @@ type recoverydata struct {
...
-		return own, nil
+		// check if any atx is being built
+		if cfg.PostDataDir == "" {
+			return nil, nil, nil
+		}
+		m, err := initialization.LoadMetadata(cfg.PostDataDir)
+		if err != nil {
+			return nil, nil, fmt.Errorf("read post metdata %s: %w", cfg.PostDataDir, err)
+		}
+		seed = types.ATXID(types.BytesToHash(m.CommitmentAtxId))
```
Comment (dshulyak) [inline checkpoint/recovery.go]: maybe top, prev or ref. i think seed is not in the right context

File: checkpoint/recovery.go [hunk 2]
```diff
@@ -216,65 +219,149 @@ type recoverydata struct {
...
+		logger.With().Debug("found commitment atx from metadata", seed)
+	} else {
+		seed = atxid
+		logger.With().Debug("found own atx", seed)
```
Comment (dshulyak) [inline checkpoint/recovery.go]: named fields (e.g when thing is passed to the logger without name) have to be deprecated eventually

File: checkpoint/recovery.go [hunk 3]
```diff
@@ -216,65 +219,149 @@ type recoverydata struct {
...
 	}
+	var seed types.ATXID
+	var own bool
 	if atxid == types.EmptyATXID {
-		return own, nil
+		// check if any atx is being built
+		if cfg.PostDataDir == "" {
+			return nil, nil, nil
+		}
+		m, err := initialization.LoadMetadata(cfg.PostDataDir)
```
Comment (dshulyak) [inline checkpoint/recovery.go]: do you want to check for positioning atx in nipost challenge? i think it becomes even more error prone , but technically same problem exists there

File: checkpoint/recovery.go [hunk 4]
```diff
@@ -347,30 +441,35 @@ func recoverFromLocalFile(
...
-			for _, atx := range own.preserve {
+		if len(deps) > 0 {
+			logger.With().Info("preserving deps for own atx",
+				log.Context(ctx),
+				log.Int("num atx", len(deps)),
+			)
+			for i, atx := range deps {
 				if err = atxs.Add(tx, atx); err != nil {
 					return fmt.Errorf("preserve atx %s: %w", atx.ID().String(), err)
 				}
```
Comment (dshulyak) [inline checkpoint/recovery.go]: i think this object (and poet proof) should pass validation before saving them to database, as it is locally configurable, nodes may save this and then advertise it to the whole cluster which will make them eventually disconnected

File: checkpoint/recovery.go (hunk 1 above)
Comment (fasmat) [reply checkpoint/recovery.go]: It is part of the seed for post, but I would also prefer to just call it `commitment`.

File: checkpoint/recovery.go (hunk 4 above)
Comment (countvonzero) [reply checkpoint/recovery.go]: yes. i have a local change that does exactly that... but testing in checkpoint test is trickier. still working on that

File: checkpoint/recovery.go (hunk 1 above)
Comment (countvonzero) [reply checkpoint/recovery.go]: @fasmat seed is in the context of using this atx as a starting point for expanding dependencies. will change it to ref as dmitry suggested

File: checkpoint/recovery.go (hunk 4 above)
Comment (dshulyak) [reply checkpoint/recovery.go]: i realized that under current constraints (e.g snapshotted atxs) this solution works only for the very first checkpoint, or more specifically for those that started initialization before first checkpoint. in all other cases snapshotted atx will be on the path

Comment (countvonzero) [reply checkpoint/recovery.go]: not necessarily. it will still work for
- own atx deps are traced to the current snapshot
- miner starts initializing in btwn snapshots and uses a commitment atx A not in old snapshot. this is likely as the algo tries to pick the max height. the only caveat is A cannot reference anything in the old snapshot. 

maybe when picking commitment/positioning atxes we should prioritize Golden() == false? but that should already happen by choosing max height.

Comment (dshulyak) [reply checkpoint/recovery.go]: i think anything that algo picks will have snapshotted atx on the path because of the reference to positioning

Comment (countvonzero) [reply checkpoint/recovery.go]: but prevAtx is a problem

Comment (dshulyak) [reply checkpoint/recovery.go]: yes that too. but positioning will be actually either the highest from the snapshot or one on top of that, depending on the timing

Comment (countvonzero) [reply checkpoint/recovery.go]: @dshulyak 
what do you think that we change the algo for a miner's first atx after snapshot to
- use golden atx for positioning/commitment atxes
- use empty atx for prev 

or alternatively we keep ALL historical snapshot atxes

should bring this up for discussion in research.

**Comment by countvonzero** [comment]
> what is the expected behavior when format has changed and user preserved own atx? will it crush its own node when it tries to encode atx? will it be disconnected from whole network because it will serve malformed data repeatedly?

yes. i have a follow up change to validate own atxs before preserving it. if it fails, then it will give up.
ideally we will also just set the flag to false when we know a priori that format is incompatible

File: checkpoint/recovery.go (hunk 4 above)
Comment (dshulyak) [reply checkpoint/recovery.go]: > use golden atx for positioning/commitment atxes

blocks in the next epoch will have lower height than the ballots from the previous one. it will definitely create more mess, not sure if it is worth it to analyze. for instance it will be one more consideration for concentration attack

> or alternatively we keep ALL historical snapshot atxes

what is the problem with configuring 4-10 epochs, as discussed yesterday? continuous post init should be supported and optimal from maximizing rewards. it is also fine to just keep all atx's metadata in the snapshot, to avoid such problems completelyy

Comment (countvonzero) [reply checkpoint/recovery.go]: > what is the problem with configuring 4-10 epochs, as discussed yesterday?

i don't see a problem. i assume @fasmat is working on that separately.

Comment (dshulyak) [reply checkpoint/recovery.go]: i don't understand the advantage of doing `preserve-own-atx` given it limited utility. especially if it requires making some adjustments in consensus

Comment (countvonzero) [reply checkpoint/recovery.go]: > i don't understand the advantage of doing preserve-own-atx given it limited utility. especially if it requires making some adjustments in consensus

are you suggesting to forgo support for inflight atxs (those already built but not in snapshot, and those still initializing)?
i am happy to delete the whole codepath actually, but the UX is going to be bad.

Comment (dshulyak) [reply checkpoint/recovery.go]: > those already built but not in snapshot

i forgot that we may be forced to do the snapshot close to cycle gap or shortly after, network should be literally dying to justify that. thanks, i remember previous chat now

so if 4-10 previous atxs are already in snapshot, then every smesher should succeed with `preserve-own-atx`?

Comment (countvonzero) [reply checkpoint/recovery.go]: > so if 4-10 previous atxs are already in snapshot, then every smesher should succeed with `preserve-own-atx`

yes, unless this miner start initializing 4 epochs ago. (assuming miners publish every epoch)
if it takes > 4 epochs to initialize their storage, we need to warn users about this possibility of being excluded from checkpoint snapshot.

File: checkpoint/recovery.go (hunk 3 above)
Comment (countvonzero) [reply checkpoint/recovery.go]: added. thanks

**Comment by countvonzero** [comment]
bors try

File: common/types/checkpoint.go [hunk 5]
```diff
@@ -1,17 +1,17 @@
-package checkpoint
+package types
```
Comment (fasmat) [inline common/types/checkpoint.go:1]: We tend to put all type definitions in this package, but I'm not a fan of separating types from the implementation. I know they are also needed by systests, but the issue here is that the implementation defines the API, when it should be the other way round.

EDIT: I don't think this needs to be addressed in this PR but imo we should reconsider the way we organize our codebase.

File: checkpoint/util.go [hunk 6]
```diff
@@ -104,7 +110,13 @@ func httpToLocalFile(ctx context.Context, resource *url.URL, fs afero.Fs, dst st
...
-		return fmt.Errorf("http get bootstrap file: %w", err)
+		if _, ok := err.(*url.Error); ok {
+			return ErrCheckpointNotFound
+		}
+		return fmt.Errorf("http get recovery file: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return ErrCheckpointNotFound
 	}
 	defer resp.Body.Close()
```
Comment (fasmat) [inline checkpoint/util.go]: `errors.As` is a more thorough check than casting the error. Also the body should be closed if `Do` didn't return an error, even if the status code is not `http.StatusOK`.

```golang
	resp, err := (&http.Client{}).Do(req)
	urlErr := &url.Error{}
	switch {
	case errors.As(err, &urlErr):
		return ErrCheckpointNotFound
	case err != nil:
		return fmt.Errorf("http get recovery file: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNotFound {
		return ErrCheckpointNotFound
	}
```

File: node/bad_peer_test.go [hunk 7]
```diff
@@ -56,11 +56,11 @@ func TestPeerDisconnectForMessageResultValidationReject(t *testing.T) {
 	})
 	g := errgroup.Group{}
 	g.Go(func() error {
-		return app1.Start(ctx)
+		return app1.Start(ctx, nil)
```
Comment (fasmat) [inline node/bad_peer_test.go]: Suggestion: use a functional option like `WithCheckpoint` to avoid having to add `nil` to every call of `Start`.

EDIT: or remove the parameter completely (see below)

File: node/node.go [hunk 8]
```diff
@@ -150,15 +151,17 @@ func GetCommand() *cobra.Command {
 					return fmt.Errorf("could not retrieve identity: %w", err)
 				}
 
-				if err := app.LoadCheckpoint(ctx); err != nil {
+				preserve, err := app.LoadCheckpoint(ctx)
+				if err != nil {
 					return err
 				}
```
Comment (fasmat) [inline node/node.go]: Alternatively to the functional option, `app.LoadCheckpoint` could just set a field that `app.Start` can access later.

File: node/node.go [hunk 9]
```diff
@@ -1394,10 +1367,57 @@ func (app *App) Start(ctx context.Context) error {
...
+					app.log.With().Error("failed to preserve atx after checkpoint",
+						log.Inline(vatx),
+						log.Err(err),
+					)
+				} else {
+					app.log.With().Info("preserved atx after checkpoint",
+						log.Inline(vatx),
+					)
+				}
+			}
```
Comment (fasmat) [inline node/node.go]: Deep nesting makes the code hard to follow imo, maybe `continue` would help?

```suggestion
		for _, vatx := range preserve.Deps {
			encoded, err := codec.Encode(vatx)
			if err != nil {
				app.log.With().Error("failed to encode atx after checkpoint",
					log.Inline(vatx),
					log.Err(err),
				)
				continue
			}
			if err := app.atxHandler.HandleAtxData(ctx, p2p.NoPeer, encoded); err != nil {
				app.log.With().Error("failed to preserve atx after checkpoint",
					log.Inline(vatx),
					log.Err(err),
				)
				continue
			}
			app.log.With().Info("preserved atx after checkpoint",
				log.Inline(vatx),
			)
		}
```

File: node/node.go [hunk 10]
```diff
@@ -1394,10 +1367,57 @@ func (app *App) Start(ctx context.Context) error {
 			app.host.ID().String(), app.Config.Genesis.GenesisID().ShortString())
 	}
 
-	if err := app.startServices(ctx, appErr); err != nil {
+	if err := app.startServices(ctx); err != nil {
 		return err
 	}
 
+	// need post verifying service to start first
+	if preserve != nil {
```
Comment (fasmat) [inline node/node.go]: I feel like the whole block inside the if could be its own method `app.restore(preserve)`.

Also I think `chkptData` or similar could be a better name for `preserve`.

File: node/node.go [hunk 11]
```diff
@@ -1394,10 +1367,57 @@ func (app *App) Start(ctx context.Context) error {
...
+						log.Err(err),
+					)
+				} else {
+					app.log.With().Info("preserved poet proof after checkpoint",
+						log.Stringer("atx id", preserve.Deps[i].ID()),
+						log.String("poet proof ref", preserve.Deps[i].GetPoetProofRef().ShortString()),
+					)
+				}
+			}
+		}
```
Comment (fasmat) [inline node/node.go]: Same as the other, `continue` helps to prevent deep nesting.

```suggestion
		for i, poetProof := range preserve.Proofs {
			encoded, err := codec.Encode(poetProof)
			if err != nil {
				app.log.With().Error("failed to encode poet proof after checkpoint",
					log.Stringer("atx id", preserve.Deps[i].ID()),
					log.Object("poet proof", poetProof),
					log.Err(err),
				)
				continue
			}
			if err := app.poetDb.ValidateAndStoreMsg(ctx, p2p.NoPeer, encoded); err != nil {
				app.log.With().Error("failed to preserve poet proof after checkpoint",
					log.Stringer("atx id", preserve.Deps[i].ID()),
					log.String("poet proof ref", preserve.Deps[i].GetPoetProofRef().ShortString()),
					log.Err(err),
				)
				continue
			}
			app.log.With().Info("preserved poet proof after checkpoint",
				log.Stringer("atx id", preserve.Deps[i].ID()),
				log.String("poet proof ref", preserve.Deps[i].GetPoetProofRef().ShortString()),
			)
		}
```

File: systest/tests/checkpoint_test.go [hunk 12]
```diff
@@ -102,6 +103,43 @@ func TestCheckpoint(t *testing.T) {
...
+	endpoint := fmt.Sprintf("http://%s:%d", ip, 80)
+	updateUrl := fmt.Sprintf("%s/updateCheckpoint", endpoint)
+	tctx.Log.Infow("submit checkpoint data", "update url", updateUrl)
+	require.NoError(t, updateCheckpointServer(tctx, updateUrl, newSnapshot.Bytes()))
+
+	queryUrl := fmt.Sprintf("%s/checkpoint", endpoint)
+	tctx.Log.Debugw("query checkpoint data", "query url", queryUrl)
+	data, err := query(tctx, queryUrl)
+	require.NoError(t, err)
+	require.True(t, bytes.Equal(newSnapshot.Bytes(), data))
```
Comment (fasmat) [inline systest/tests/checkpoint_test.go]: This helps to see what the difference is should the assertion fail:

```suggestion
	require.Equal(t, newSnapshot.String(), string(data))
```

Also imo instead of serializing the expected data to json, it would be better to deserialize the received bytes to an object and compare those two. Arguably these two are the same: `{"field1":1,"field2":2}` and `{"field2":2,"field1":1}` but comparing them as strings will cause the assertion to fail. Same with nested objects and order of elements in arrays.

File: checkpoint/recovery_test.go [hunk 13]
```diff
@@ -195,24 +145,30 @@ func TestRecoverFromHttp(t *testing.T) {
...
 			db := sql.InMemory()
-			newdb, err := checkpoint.RecoverWithDb(ctx, logtest.New(t), db, fs, cfg, types.NodeID{2, 3, 4}, tc.uri, types.LayerID(recoverLayer))
-			if len(tc.expErr) > 0 {
-				require.ErrorContains(t, err, tc.expErr)
+			preserve, err := checkpoint.RecoverWithDb(ctx, logtest.New(t), db, fs, cfg)
+			if tc.expErr != nil {
+				require.ErrorIs(t, err, tc.expErr)
 				return
 			}
 			require.NoError(t, err)
```
Comment (fasmat) [inline checkpoint/recovery_test.go:165]: No need to be careful. `errors.Is` can handle `nil`:

```suggestion
			require.ErrorIs(t, err, tc.expErr)
```

https://go.dev/play/p/yZfoyOh7n2g

File: checkpoint/recovery_test.go [hunk 14]
```diff
@@ -362,172 +393,383 @@ func TestRecover_OwnAtxInCheckpoint(t *testing.T) {
...
+	validateAndPreserveData(t, olddb, vatxs[1:], proofs[1:])
+	// the proofs are not valid, but save them anyway for the purpose of testing
+	for i, vatx := range vatxs {
+		if i == 0 {
+			continue
+		}
+		encoded, err := codec.Encode(proofs[i])
+		require.NoError(t, err)
+		require.NoError(t, poets.Add(olddb, types.PoetProofRef(vatx.GetPoetProofRef()), encoded, proofs[i].PoetServiceID, proofs[i].RoundID))
+	}
```
Comment (fasmat) [inline checkpoint/recovery_test.go]: Ranging over proofs instead of vatxs simplifies the code a bit

```suggestion
	for i, proof := range proofs[1:] {
		encoded, err := codec.Encode(proof)
		require.NoError(t, err)
		require.NoError(t, poets.Add(olddb, types.PoetProofRef(vatx[i+1].GetPoetProofRef()), encoded, proof.PoetServiceID, proof.RoundID))
	}
```

File: checkpoint/recovery.go [hunk 15]
```diff
@@ -156,30 +85,33 @@ func copyToLocalFile(ctx context.Context, logger log.Log, fs afero.Fs, dataDir,
...
+		if errors.Is(err, ErrCheckpointNotFound) {
+			logger.With().Info("no checkpoint file available. not recovering",
+				log.String("uri", cfg.Uri),
+			)
+			return nil, nil
 		}
+		return nil, err
 	}
-	return nil
+	return preserve, nil
```
Comment (fasmat) [inline checkpoint/recovery.go]: switching makes the error handling a bit easier to understand I think:

```golang
	preserve, err := RecoverWithDb(ctx, logger, db, fs, cfg)
	switch {
	case errors.Is(err, ErrCheckpointNotFound):
		logger.With().Info("no checkpoint file available. not recovering",
			log.String("uri", cfg.Uri),
		)
		return nil, nil
	case err != nil:
		return nil, err
	default:
		return preserve, nil
	}
```

File: node/node.go (hunk 10 above)
Comment (countvonzero) [reply node/node.go]: but these are not checkpointData. at this point the checkpoint data is already saved in db.
here `preserve` means the data we want to preserve from the previous db, and is trying to save them to the new db via the gossip validation path.

File: systest/tests/checkpoint_test.go (hunk 12 above)
Comment (countvonzero) [reply systest/tests/checkpoint_test.go]: in the systest i am trying hard to be black box (not knowing the data transpired btwn nodes.
here all i care is the data is the same as the one i just updated.

File: checkpoint/recovery_test.go (hunk 13 above)
Comment (countvonzero) [reply checkpoint/recovery_test.go:165]: i did this intentionally because i am returning immediately when there is error,

File: checkpoint/recovery_test.go (hunk 14 above)
Comment (countvonzero) [reply checkpoint/recovery_test.go]: but i think the suggested approach is more error-prone because it manipulating the index. here skipping 0 is very explicit. i will range over proofs instead but keeping continue.

File: checkpoint/recovery.go (hunk 15 above)
Comment (countvonzero) [reply checkpoint/recovery.go]: for me if/else block is easier to read.....but ok.

File: node/node.go (hunk 8 above)
Comment (countvonzero) [reply node/node.go]: done. thanks for the suggestion.

File: node/node.go (hunk 11 above)
Comment (countvonzero) [reply node/node.go]: moved to a separate function and used continue.

**Comment by countvonzero** [comment]
bors try

File: systest/tests/checkpoint_test.go (hunk 12 above)
Comment (fasmat) [reply systest/tests/checkpoint_test.go]: The problem is that if that assertion fails it doesn't display way:

```
FAIL TestName
     Reason: Should be true
```

Which is especially hard to fix later should the test become flaky and a failing test doesn't say why it failed, for this I think using strings and `require.Equal` would be helpful over `require.True`.

The second (less important) issue is that both values can be different but arguably the test should still pass in my example above.

**Review by fasmat — APPROVED** [review]

File: systest/tests/checkpoint_test.go (hunk 12 above)
Comment (countvonzero) [reply systest/tests/checkpoint_test.go]: the only reason i serialize data into json is so that i can manipulate the snapshot such that half of the miners need to preserve their own atxs after checkpoint. i'll deserialize the byte to match the content.

**Comment by countvonzero** [comment]
bors merge
